quantum/process_keycode/process_auto_shift.c

```diff
@@ -33,23 +42,44 @@ void autoshift_timer_report(void) {
     send_string((const char *)display);
 }
 
-void autoshift_on(uint16_t keycode) {
-    autoshift_time    = timer_read();
+bool autoshift_on(uint16_t keycode) {
+    if (!autoshift_enabled) { return true; }
+#    ifndef AUTO_SHIFT_MODIFIERS
+    if (get_mods()) {
+        return true;
+    }
+#    endif
+    autoshift_time = timer_read();
     autoshift_lastkey = keycode;
+
+    // We need some extra handling here for OSL edge cases
+#    if !defined(NO_ACTION_ONESHOT) && !defined(NO_ACTION_TAPPING)
+    clear_oneshot_layer_state(ONESHOT_OTHER_KEY_PRESSED);
+#    endif
+    return false;
 }
 
 void autoshift_flush(void) {
-    if (autoshift_lastkey != KC_NO) {
+    if (autoshift_lastkey != -1) {
         uint16_t elapsed = timer_elapsed(autoshift_time);
 
+#    ifdef PERMISSIVE_HOLD
```

Comment:
```suggestion
#    if defined(PERMISSIVE_HOLD) || defined(PERMISSIVE_HOLD_PER_KEY)
```

To account for pending [PR #7994](https://github.com/qmk/qmk_firmware/pull/7994)

Reply:
I'm thinking I'll move `old_weak_mods` outside of this `#if` so that the one following this can just be the `||`. This one will need to be two separate `#if`s as if `PERMISSIVE_HOLD_PER_KEY` is defined it will also need to check `get_permissive_hold()` before (possibly) setting weak mods.

Reply:
Yeah that's probably a better approach in all honesty.